hbase-server/src/main/java/org/apache/hadoop/hbase/quotas/QuotaCache.java

```diff
@@ -160,6 +172,27 @@ protected boolean isExceedThrottleQuotaEnabled() {
     return exceedThrottleQuotaEnabled;
   }
 
+  /**
+   * Applies a request attribute user override if available, otherwise returns the UGI's short
+   * username
+   * @param ugi The request's UserGroupInformation
+   */
+  private String getQuotaUserName(final UserGroupInformation ugi) {
+    if (userOverrideRequestAttributeKey == null) {
+      return ugi.getShortUserName();
+    }
+
+    Optional<RpcCall> rpcCall = RpcServer.getCurrentCall();
+    if (
+      rpcCall.isPresent()
+        && rpcCall.get().getRequestAttributes().containsKey(userOverrideRequestAttributeKey)
```

Comment:
I might recommend just looping the raw `rpcCall.get().getHeader().getAttributes()` here. For a small N (as i expect attributes to be), it is typically faster to iterate the list than to lookup in a map. This is especially true if we have to actually build the map as well (as we are likely to here). So in a vacuum, this call loops all the attributes and populates a map (which involves hashcode and equals for each, not to mention converting from ByteString to byte[] and string). Then we do a lookup, which involves another hashcode and equals. On the other hand, a loop can exit out early if a match is found and only requires equals.

So if we planned to do many operations on the map, it might start to pay off to do those conversions. But if we are just looking for a single key in an empty or very small list, it'd be cheaper to skip all of that.

Reply:
yeah that makes sense, seems like a reasonable optimization. should we bake that into an `RpcCall#getRequestAttribute(String key)` method to help this pattern be standard or is that overkill?

Reply:
Yea, let's do that. We can have it pull from the map if it's non-null, and otherwise loop. Just add javadoc describing the difference between that and getRequestAttributes() and when to use each.